sdorfehs.1
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2 of the License, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU General Public License along with
.\" this program; if not, write to the Free Software Foundation, Inc., 59 Temple
.\" Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.\" This manpage is written using the mdoc macro language.
.\" Examples of formatters which support mdoc are groff[1] and mandoc[2].
.\" An mdoc language reference is available[3].
.\"
.\" [1] https://gnu.org/software/groff/
.\" [2] http://mdocml.bsd.lv/
.\" [3] http://mdocml.bsd.lv/mdoc.7.html
.\"
.Dd Aug 22, 2019
.Dt SDORFEHS 1
.Os
.Sh NAME
.Nm sdorfehs
.Nd window manager without mouse dependency
.Sh SYNOPSIS
.Nm
.Op Fl hv
.Nm
.Op Fl d Ar dpy
.Op Fl s Ar num
.Op Fl f Ar file
.Nm
.Op Fl d Ar dpy
.Op Fl s Ar num
.Op Fl i
.Fl c Ar command Op Fl c Ar command ...
.Sh DESCRIPTION
.Nm
is a window manager without fat library dependencies, fancy graphics or
dependence on a mouse.
.Pp
The screen can be split into non-overlapping frames.
All windows are kept maximized inside their frames.
.Pp
All interaction with the window manager is done through keystrokes.
.Nm
has a prefix map to minimize key clobbering.
.Pp
The options are as follows:
.Bl -tag -width Bs
.It Fl c Ar command
Send
.Nm
a command.
There must be a
.Nm
instance running as window manager for the given display/screen for this to
work.
Do not forget to quote the command if it contains spaces.
For example:
.Pp
.Dl Nm Fl c Qq Ar "echo hello world"
.It Fl d Ar display
Set the X display to use or send commands to.
.It Fl f Ar filename
Specify an alternate configuration file.
If this is not given,
.Nm
will try
.Pa $HOME/.config/sdorfehs/config
and execute each command when starting up.
.It Fl h
Show summary of options.
.It Fl i
Execute commands given with
.Fl c
in interactive mode.
That means it will behave exactly as if called with
.Ic C\-a \&:
like prompting for missing arguments and things like that.
.It Fl s Ar number
Only use the specified screen.
.El
.Sh KEY BINDINGS
To avoid conflicts with other programs, all default
.Nm
key bindings start with an escape key which, by default, is
.Ic C\-a
(Control\-a).
Some important default key bindings:
.Bl -tag -width Ds
.It Ic C\-a \&?
Show key bindings
.It Ic C\-a a
Send a Control\-a to the current window
.It Ic C\-a c
Start an X terminal
.It Ic C\-a n
Switch to next window
.It Ic C\-a p
Switch to previous window
.It Ic C\-a 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
Switch to window number 0 | 1 | 2 | ...
.It Ic C\-a k
Close the current window
.It Ic C\-a K
XKill the current application
.It Ic C\-a s | S
Split the current frame into two vertical | horizontal ones
.It Ic C\-a Tab | Left | Up | Right | Down
Switch to the next | left | top | right | bottom frame.
.It Ic C\-a Q
Make the current frame the only one
.It Ic C\-a \&:
Execute a
.Nm
command
.It Ic C\-a F1 | F2 | F3 | F4 | F5 | F6 | F7 | F8 | F9 | F10 | F11 | F12
Switch to vscreen number 0 | 1 | 2 | ...
.El
.Pp
Further default key bindings can be found in parentheses after the
commands in the next section.
.Sh COMMANDS AND DEFAULT ALIASES
.Bl -tag -width Ds
.It Ic abort
.Pq Ic C\-a C\-g
Do nothing and that successfully.
.Po
Useful if you pressed
.Ic C\-a
in error
.Pc .
.It Ic addhook Ar event command
Add a hook: run
.Ar command
whenever
.Ar event
is called.
Possible events are:
.Bl -tag -width Ds
.It Cm deletewindow
Run after a window is withdrawn.
.It Cm key
Run whenever a top level key is pressed (by default
.Ic C\-a ) .
.It Cm newwindow
Run after a new window is mapped.
.It Cm quit
Run before exiting
.Nm .
.It Cm restart
Run before restarting
.Nm .
.It Cm switchframe
Run after a frame actually switched, but before the window in it is
focused.
.It Cm switchscreen
Run when the user switches to a different screen.
.It Cm switchvscreen
Run when the user switches to a different virtual screen.
.It Cm switchwin
Run after a new window is selected.
(With dedication, it may already be inactive again, if it was put into
another frame)
.El
.It Ic alias Ar alias command
Add
.Ar alias
as new way to call
.Ar command .
.It Ic bind Ar key command
alias for
.Qq Ic definekey Li root Ar key command
.It Ic banish
.Pq Ic C\-a b
Banish the rat cursor to the lower right corner of the screen.
.It Ic banishrel
Banish the rat cursor to the lower right corner of the current window.
If there isn't a window in the current frame, it banishes the rat cursor
to the lower right corner of the frame.
.It Ic chdir Op Ar directory
If the optional argument is given, change the current directory of
.Nm
to
.Ar directory .
If nothing is given, change it to the value of the environment variable
.Qq Ev HOME .
.It Ic clrunmanaged
Clears the unmanaged window list.
.It Ic cnext
Like
.Ic next
but switch to the next window with another resource class than the
current one.
(That means the next window belonging to another type of application
than the current one.)
.It Ic cprev
Like
.Ic prev
but switch to the previous window with another resource class than the
current one.
(That means the previous window belonging to another type of application
than the current one.)
.It Ic colon Ar sdorfehs\-command Pq Ic C\-a \&:
Execute
.Ar sdorfehs\-command
interactively.
(i.e. ask for possible missing arguments.)
.It Ic cother
Like
.Ic other
but switch to the window of the current vscreen that was last accessed and
has another resource class but is not currently visible.
.It Ic curframe Pq Ic C\-a F
Show a bar marking the current frame.
.It Ic definekey Ar keymap key command
Add a new key binding in
.Ar keymap
for
.Ar key
to execute
.Ar command .
Default keymaps are
.Li top
normally only containing
.Ic C\-a ,
which reads a key from
.Li root ,
containing all the normal commands.
.Pp
Note that you have to describe ":" by "colon", "!" by "exclam" and so on.
If you cannot guess a name of a key, try either
.Ic C\-a Ar key
and look at the error message, or try
.Ic :describekey Li root
and pressing the key.
.It Ic dedicate Op Cm 0 | 1
Consider the current frame dedicated/chaste
.Pq Cm 1
or promiscuous
.Pq Cm 0 .
.Pp
A dedicated frame will not accept new windows.
When new windows are to be focused, they will be opened in a non-dedicated
frame instead.
.Pp
If no argument is given, toggle the current dedicateness.
By default no windows are dedicated.
.It Ic delete Pq Ic C\-a k
Close the current window.
.It Ic delkmap Ar keymap
Deletes the keymap named
.Ar keymap ,
that was generated with
.Ic newkmap .
The keymaps
.Li top
(or whatever was specified by
.Ic set Ar topkmap )
and
.Li root
cannot be deleted.
.It Ic describekey Ar keymap
Grab the next key.
Similar to
.Ic readkey ,
.Ic describekey
shows only the command in
.Ar keymap ,
that would be executed by
.Ic readkey .
.It Ic echo Ar text
Show
.Ar text
as
.Nm
message.
.It Ic escape Ar key
Update the default escape key to
.Ar key .
.Pp
Strictly speaking it updates the
.Ic readkey Li root
command in the keymap
.Li top
to
.Ar key ,
the
.Ic other
binding
in
.Li root
to
.Ar key ,
and
.Ic meta
binding in
.Li root
to
.Ar key
without modifiers or
.Ic "C\-" Ns Ar key
if
.Ar key
has no modifiers.
(If
.Ic set Ar topkmap
was called with an argument other than
.Ar top
that will be used instead of
.Ar top . )
.It Ic exchangedown Pq Ic C\-a C\-Down
Exchange the window in the current frame with the window in the frame
below the current frame.
.It Ic exchangeleft Pq Ic C\-a C\-Left
Exchange the window in the current frame with the window in the frame
left of the current frame.
.It Ic exchangeright Pq Ic C\-a C\-Right
Exchange the window in the current frame with the window in the frame
right of the current frame.
.It Ic exchangeup Pq Ic C\-a C\-Up
Exchange the window in the current frame with the window in the frame
above the current frame.
.It Ic exec Ar shell\-command Pq Ic C\-a \&!
Spawn a shell executing
.Ar shell\-command .
.It Ic execa Ar shell\-command
Spawn a shell executing
.Ar shell\-command ,
without remembering the current frame, so that _NET_WM_PID declaring
programs will be placed into the frame active when they open a window
instead of the frame active when
.Nm
gets this command.
.It Ic execf Ar frame shell\-command
Spawn a shell executing
.Ar shell\-command ,
showing _NET_WM_PID supporting programs in the given frame instead of
the frame selected when this program is run.
.It Ic execv Ar vscreen shell\-command
Spawn a shell within a specific vscreen
.Ar vscreen ,
executing
.Ar shell\-command
as in the 'exec' command (i.e., using "sh -c").  This is not likely to work
well with 'winaddcurvscreen' set to 1 anywhere in the configuration file.
.It Ic execw Ar shell\-command
Spawn a shell executing
.Ar shell\-command ,
but wait for it to finish.  This should behave the same as
.Ar exec
but will wait for completion.  Use timeout(1) if it might hang, or
.Nm
will hang itself waiting for completion.  Use the 'exec' shell builtin as the
first word in the command string to get similar behavior to
.Ar execa
owing to the forked pid remaining stable and known to
.Nm .
.It Ic fdump Op Ar screenno
Output the defining data for all frames of the current screen, or
for screen number
.Ar screenno
if this is specified.
.It Ic focus Pq Ic C\-a Tab
Focus the next frame.
.It Ic focuslast
Switch to the last selected focus.
.It Ic focusleft Pq Ic C\-a Left
Switch to the frame to the left of the current one.
.It Ic focusdown Pq Ic C\-a Down
Switch to the frame beneath the current one.
.It Ic focusright Pq Ic C\-a Right
Switch to the frame to the right of the current one.
.It Ic focusprev
Focus the previous frame.
.It Ic focusup Pq Ic C\-a Up
Switch to the frame above the current one.
.It Ic frestore Ar frames
Replace the current frames with the ones specified in
.Ar frames
in the format as generated by
.Ic fdump .
.It Ic fselect Oo Ar frameno Oc Pq Ic C\-a f
If an argument is supplied, switch to a frame given by number
.Ar frameno .
.Pp
If no argument is given, show a frame selector in each frame and wait for
a key to be pressed.
If the key matches an existing frame selector, this frame gets focused.
.Pp
Frame selectors are by default the numbers starting with zero, but they
can be changed by
.Ic set Ns
ting
.Ar framesels .
.It Ic getenv Ar variable
Output the value of the environment variable
.Ar variable .
.It Ic getsel
Paste the current X Selection into the current window.
.It Ic gravity Op Cm nw | w | sw | n | c | s | ne | e | se
Change how in its frame the current window is aligned.
.It Ic help Op Ar keymap
If the optional parameter
.Ar keymap
is given, list all keybindings in this keymap, otherwise list all key
bindings in keymap
.Li root .
.It Ic hsplit Oo Ar l Ns Li / Ns Ar p | Oo Li \- Oc Ns Ar pixels Oc Pq Ic C\-a S
Split the current frame into left frame and a right frame.
If no parameter is given, split in halves.
If two numbers separated by a slash
.Pq Ql "/"
are given, the left one is
.Ar l
times the
.Ar p Ns
th
part and the right one
.Pq Ar p Li \- Ar l
times the
.Ar p Ns
th
part of the prior width.
Otherwise the right half is
.Ar pixels
wide or the left one is
.Ar pixels
wide, depending whether there is
.Ql \-
in front of the number or not.
.It Ic inext
Like
.Ic next
but switch to the next window with the same resource class as the
current one.
(That means the next window belonging to the same application
as the current one.)
.It Ic info Pq Ic C\-a i
Output the current the width, height, window number and window name of
the current window.
.Po
What name means is chosen by
.Dq Ic set Ar winname .
.Pc
.It Ic iprev
Like
.Ic prev
but switch to the previous window with the same resource class as the
current one.
(That means the previous window belonging to the same application as the
current one.)
.It Ic iother
Like
.Ic other
but switch to the window of the current vscreen that was last accessed and
has the same resource class but is not currently visible.
.It Ic kill Pq Ic C\-a K
Close the X-connection of the X-client responsible for the current window.
.It Ic lastmsg Pq Ic C\-a m
Reshow the last message.
.It Ic link Ar key Op Ar keymap
Do what
.Ar key
is bound to in the keymap
.Ar keymap
if supplied.
Otherwise what
.Ar key
is bound to in keymap
.Li root .
.It Ic listhook Ar event
List all commands specified with
.Ic addhook
to be executed when even
.Ar event
occurs.
.It Ic meta Oo Ar key Oc Pq Ic C\-a t
Send the escape key (that which normally is
.Ic C\-a )
to the current window.
If a
.Ar key
is specified, this is sent instead.
Note that some applications by default ignore the synthetic key that is
sent using this command as it is considered a security hole.
xterm is one such application.
.It Ic newkmap Ar keymap
Generate a new keymap named
.Ar keymap .
This keymap can be used to add new key-command mappings to it with
.Ic definekey
and can be called with
.Ic readkey .
.It Ic next Pq Ic C\-a Return | C\-a n | C\-a space
Switch to the next window in the current vscreen.
.It Ic nextscreen Pq Ic C\-a N
Switch to the next screen. (If you have multiple physical ones.)
.It Ic number Ar new Op Ar old
Give the number
.Ar new
to the window with the number
.Ar old
or the current window.
.It Ic only Pq Ic C\-a Q
Remove all frames on the current screen except the current frame and
maximize this one to the size of the whole screen.
.It Ic other Pq Ic C\-a C\-a
Switch to the window of the current vscreen that was last
accessed but is not currently visible.
.It Ic prev Pq Ic C\-a p
Switch to the previous window in the current vscreen.
.It Ic prevscreen Pq Ic C\-a P
Switch to the previous screen. (If you have multiple physical ones.)
.It Ic prompt Op Ar prompt
.Nm
will ask the user for input, showing
.Ar prompt
(or a single colon, if no argument is given) and output the input the
user has made.
Note that this command probably does not make much sense in interactive
mode.
.It Ic putsel Ar x\-selection
Replace the X selection with the text
.Ar x\-selection .
It can be inserted into the current window with
.Ic getsel .
.It Ic quit
Quit
.Nm .
.It Ic ratinfo
Display the x y coordinates of the rat cursor relative to the screen.
.It Ic ratrelinfo
Display the x y coordinates of the rat cursor relative to the current
window or current frame if no window is focused
.It Ic ratwarp Ar x y
Move the rat cursor to the position
.Ar ( x , y ) .
.It Ic ratrelwarp Ar deltax deltay
Move the rat cursor to
.Ar ( deltax , deltay ) ,
relative to the current position.
.It Ic ratclick Op Ar button
Simulate a rat click with
.Ar button
(button 1=left button if none given).
.It Ic rathold Cm ( up | down ) Op Ar button
Simulate pressing|releasing rat button
.Ar button
(1=left button if none given).
.It Ic readkey Ar keymap
Grab the next key pressed, and execute the command associated to this key
in
.Ar keymap .
To show it is waiting for a key,
.Nm
will change the rat cursor to a square if
.Va waitcursor
is set.
This command is perhaps best described with its usage in the default
configuration: by pressing
.Ic C\-a ,
which is the only key in the keymap
top ,
the command
.Qq Ic readkey Ar root
is executed.
The next key then executes the command in keymap
.Li root
belonging to
this command.
.It Ic redisplay Pq Ic C\-a l
Extend the current window to the whole size of its current frame and
redisplay it.
(Useful to redisplay normal windows or bring transient windows to the
full size of the frame as only normal windows are maximized by
.Nm )
.It Ic redo Pq Ic C\-a U
Revert the last
.Ic undo
of frame changes.
.It Ic remhook Ar event command
Remove command
.Ar command
from the list of commands to be called when event
.Ar event
is hit.
(The command has to specified, as an event can have multiple commands
attached to it.)
Use
.Qq Ic listhook Ar hook
to get a list of all attached commands.
.It Ic remove Pq Ic C\-a R
Remove the current frame and extend some frames around to fill the
remaining gap.
.It Ic resize Oo Ar deltax deltay Oc Pq Ic C\-a r
If
.Ar deltax
and
.Ar deltay
are supplied, resize the current frame by that (i.e. move the bottom
right corner by the given offsets and then move this frame and resize
adjacent frames to make the frames fill the whole screen again.)
.Pp
If in interactive mode no arguments are supplied, resize the current
frame interactively:
.Pp
.Bl -tag -offset 2n -width "C-f, Right, l" -compact
.It Ic Return
finish resizing
.It Ic C\-g , Escape
abort resizing
.It Ic C\-n , Down , j
grow vertically
.It Ic C\-p , Up , k
shrink vertically
.It Ic C\-f , Right , l
grow horizontally
.It Ic C\-b , Left , h
shrink horizontally
.It Ic s

NEWS
ratpoison NEWS --- history of user-visible changes.  -*- outline -*-

Copyright (C) 2000, 2001, 2002, 2003, 2004 Shawn Betts
Copyright (C) 2013, 2014, 2017 Jérémie Courrèges-Anglas

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.

* Changes since 1.4.9

** Feature/misc changes

*** manpage fixes

*** fix how find_frame_left, find_frame_right check for overlap

*** always show the frame indicator when switching to a different screen

*** fix bogus libhistory usage

*** fix for tmpwm and multiple screens

*** new 'onlyborder' variable
To print borders only when >=2 frames are on screen.

*** %42s-style format strings now account for UTF-8 characters

** Packaging changes

*** the debian/ subdirectory was removed (outdated)

*** a basic regress test has been added

* Changes since 1.4.8

** Incompatible changes

*** command `compat' and the various def* compat commands were removed

*** `msgwait', `rudeness', `startupmessage', `warp' are now variables
The msgwait, rudeness, startup_message and warp commands will stay
around for a few releases.  Upgrade your config file now!

** Feature/misc changes

*** XRandR support
ratpoison now uses XRandR to detect screens additions/removals at
runtime.  The Xinerama support has been removed.

*** the focus* commands can be used to navigate across screens

*** window completion match by substring, case-insensitive

*** new variable `framemsgwait'
Timeout (in seconds) for the `Current frame' message window.
If set to -1, no message is shown.

*** better `select' error handling

*** the command `set' with no argument dumps a list of all variables/values

*** unknown commands are now logged as warnings to stdout too
Should help debugging a config file.

*** various cleanups and minor bugfixes

** Packaging changes

*** ratpoison now depends on libXrandr instead of libXinerama

*** ratpoison doesn't link against libhistory by default any more
Use ./configure --enable-history to force the use of said library.

*** no ChangeLog file
Please use the git repository instead.  Savannah offers a web interface.

*** no more timestamps in compiled code
Should help reproducible builds.

* Changes since 1.4.7
This is mostly a bugfix release.

** fix cmd_sfdump with multiple screens
Broken since last release, from Joshua Leners.

** respect --enable-history=no
From Jeroen Roovers (jer@gentoo).

** fix for sloppy.c
The ugly sloppy.c had "a bug".  It may be fixed now.

** README improvements
From Jeff Abrahamson.

* Changes since 1.4.6
** More fixes and code cleanup.

** UTF-8 editing support in the input bar

** get rid of "XGetWMName failed" spam on stderr
Introduced in the previous release, sorry!

** groups can now be renumbered (`gnumber')

** contrib/rpdate.sh, fancy calendar shell script to use instead of `time'

** fix `meta', which was sending garbage and made firefox unusable

** handle errors about missing config files
- ratpoison -f missing-config-file -> error
- unreadable ~/.ratpoisonrc file -> warning

** respect --sysconfdir for the global ratpoisonrc file
Rather than a hardcoded /etc/ratpoisonrc location.

** rpws gains a new `current' command (prints the current workspace)

** new manpage, mdoc(7) formatted

** X socket closed before spawning processes
Let's save those precious file descriptors.

** DWIM windows/groups names matching
Window and group names are first looked up for exact match; in last
resort a fuzzy match is attempted.

** try to cope with unset/empty HOME env. variable

** ratpoison.el is now installed at the proper location
ie. in $prefix/share/emacs/site-lisp/

** ratpoison now restricts permissions on your history file

* Changes since 1.4.5
** *Lots* of reliability fixes in ratpoison itself
Please check out the ChangeLog for more details.

** Build process cleanup
e.g. detection of getline() was fixed, build is ok with old gcc's (2.95)

** ratpoison -c now reports errors ($? != 0)

** UTF-8 text is now printed correctly
Both with and without Xft support.

** contrib/ cleanup
Scripts should be more portable, ruby bindings work with newer ruby versions.

** New newwindow and titlechanged hooks

** Functionality was added to rpws
It can now move windows to other workspaces; workspaces can be dumped and
restored.

* Changes since 1.4.4
** bug fixes to history and xft

* Changes since 1.4.3
** add commands ratinfo, ratrelinfo, banishrel

** add history to input bar when libhistory is not used

** add settings historysize and historycompaction

** new window formatters %p, %M
prints the pid of the window and whether its a maxsize window.

** new commands grename, gother

** new bindings C-t N and C-t P
bound to nextscreen/prevscreen

** new settings, bwcolor and fwcolor
For setting the border color of focused and unfocused windows.

** netwm dialog type windows supported
These windows are treated as transients. Some Firefox popup windows
are tagged as dialog windows and should now work with ratpoison as
expected.

** new format character %M
This expands to Maxsize if the window is a maxsize window, nothing if not.

** ratpoison can use xft
If xft is found on your system ratpoison will use it unless explicitely
told not to using --without-xft.

* Changes since 1.4.2
** key event handling
top level key bindings are grabbed "asynchronously" instead of
"synchronously." This should eliminate the freezes some people have
experienced in previous versions.

* Changes since 1.4.1
Just bug fixes.

* Changes since 1.4.0
** set padding affects bar location
On OSX the bar and top of each window is hidden by the menu bar. set
padding corrects the window problem but not the bar problem. So this
change fixes that.

** new set variable framefmt
See docs for info.

** new commands swap, exchangeleft, exchangeright, exchangeup, exchangedown
See docs for info

** new commands execa and execf
See docs for info.

** default font change

sdorfehs.c
/*
 * Copyright (C) 2000, 2001, 2002, 2003, 2004 Shawn Betts <sabetts@vcn.bc.ca>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/Xproto.h>
#include <X11/cursorfont.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <sys/wait.h>
#include <ctype.h>

#include "sdorfehs.h"

static void
sighandler(int signum)
{
	kill_signalled++;
}

static void
hup_handler(int signum)
{
	hup_signalled++;
}

static void
alrm_handler(int signum)
{
	alarm_signalled++;
}

static int
handler(Display *d, XErrorEvent *e)
{
	char error_msg[100];

	if (e->request_code == X_ChangeWindowAttributes &&
	    e->error_code == BadAccess)
		errx(1, "another window manager is already running");

#ifdef IGNORE_BADWINDOW
	return 0;
#else
	if (ignore_badwindow && e->error_code == BadWindow)
		return 0;
#endif

	XGetErrorText(d, e->error_code, error_msg, sizeof(error_msg));
	warnx("X error: %s", error_msg);

	/*
	 * If there is already an error to report, replace it with this new one.
	 */
	free(rp_error_msg);
	rp_error_msg = xstrdup(error_msg);

	return 0;
}

static void
print_help(void)
{
	printf("%s %s\n", PROGNAME, VERSION);
	printf("usage: %s [-h]\n", PROGNAME);
	printf("       %s [-d dpy] [-c cmd] [-i] [-f file]\n", PROGNAME);
	exit(0);
}

static int
read_startup_files(const char *alt_rcfile)
{
	FILE *fileptr = NULL;
	char *config_dir, *filename;

	if (alt_rcfile)
		filename = strdup(alt_rcfile);
	else {
		config_dir = get_config_dir();
		filename = xsprintf("%s/config", config_dir);
		free(config_dir);
	}

	fileptr = fopen(filename, "r");
	if (fileptr == NULL && errno != ENOENT)
		warn("could not open %s", filename);
	free(filename);

	if (fileptr != NULL) {
		set_close_on_exec(fileno(fileptr));
		read_rc_file(fileptr);
		fclose(fileptr);
		return 1;
	}

	return 0;
}

/*
 * Odd that we spend so much code on making sure the silly welcome message is
 * correct.  Oh well...
 */
static void
show_welcome_message(void)
{
	rp_action *help_action;
	char *prefix, *help;
	rp_keymap *map;

	prefix = keysym_to_string(prefix_key.sym, prefix_key.state);

	map = find_keymap(ROOT_KEYMAP);

	/* Find the help key binding. */
	help_action = find_keybinding_by_action("help " ROOT_KEYMAP, map);
	if (help_action)
		help = keysym_to_string(help_action->key, help_action->state);
	else
		help = NULL;


	if (help) {
		/*
		 * A little kludge to use ? instead of `question' for the help key.
		 */
		if (!strcmp(help, "question"))
			marked_message_printf(0, 0, MESSAGE_WELCOME, prefix, "?");
		else
			marked_message_printf(0, 0, MESSAGE_WELCOME, prefix, help);

		free(help);
	} else {
		marked_message_printf(0, 0, MESSAGE_WELCOME, prefix, ":help");
	}

	free(prefix);
}

static void
init_defaults(void)
{
	unsigned long atom = 0;

	defaults.top_kmap = xstrdup(TOP_KEYMAP);

	defaults.win_gravity = NorthWestGravity;
	defaults.trans_gravity = CenterGravity;
	defaults.maxsize_gravity = CenterGravity;

	defaults.input_window_size = 200;
	defaults.window_border_width = 1;
	defaults.only_border = 1;
	defaults.bar_x_padding = 14;
	defaults.bar_y_padding = 10;
	defaults.bar_location = NorthWestGravity;
	defaults.bar_timeout = 3;
	defaults.bar_border_width = 0;
	defaults.bar_in_padding = 1;
	defaults.bar_sticky = 1;

	defaults.frame_indicator_timeout = 1;
	defaults.frame_resize_unit = 10;

	defaults.padding_left = 20;
	defaults.padding_right = 20;
	defaults.padding_top = 20;
	defaults.padding_bottom = 20;

	defaults.gap = 10;

	defaults.font_string = xstrdup(DEFAULT_XFT_FONT);

	defaults.fgcolor_string = xstrdup("#eeeeee");
	defaults.bgcolor_string = xstrdup("black");
	defaults.fwcolor_string = xstrdup("black");
	defaults.bwcolor_string = xstrdup("black");
	defaults.barbordercolor_string = xstrdup("#eeeeee");

	defaults.wait_for_key_cursor = 1;

	defaults.window_fmt = xstrdup("%n%s%t");
	defaults.info_fmt = xstrdup("(%H, %W) %n(%t)");
	defaults.frame_fmt = xstrdup("Frame %f (%Wx%H)");
	defaults.sticky_fmt = xstrdup("%t");
	defaults.resize_fmt = xstrdup("Resize frame (%Wx%H)");

	defaults.win_name = WIN_NAME_TITLE;
	defaults.startup_message = 1;
	defaults.warp = 0;
	defaults.window_list_style = STYLE_COLUMN;

	defaults.history_size = 20;
	defaults.frame_selectors = xstrdup("");
	defaults.maxundos = 20;

	get_atom(DefaultRootWindow(dpy), _net_number_of_desktops, XA_CARDINAL,
	    0, &atom, 1, NULL);
	if (atom > 0)
		defaults.vscreens = (int)atom;
	else
		defaults.vscreens = 12;

	defaults.ignore_resize_hints = 0;
	defaults.win_add_cur_vscreen = 0;
}

int
main(int argc, char *argv[])
{
	int c;
	char **cmd = NULL;
	int cmd_count = 0;
	char *display = NULL;
	int interactive = 0;
	char *alt_rcfile = NULL;
	char pid[8];

	setlocale(LC_CTYPE, "");

	if (XSupportsLocale()) {
		if (!XSetLocaleModifiers(""))
			warnx("couldn't set X locale modifiers");
	} else
		warnx("X doesn't seem to support your locale");

	/* Parse the arguments */
	myargv = argv;
	while ((c = getopt(argc, argv, "c:d:hif:")) != -1) {
		switch (c) {
		case 'c':
			cmd = xrealloc(cmd, sizeof(char *) * (cmd_count + 1));
			cmd[cmd_count++] = xstrdup(optarg);
			break;
		case 'd':
			display = optarg;
			break;
		case 'f':
			alt_rcfile = optarg;
			break;
		case 'h':
			print_help();
			break;
		case 'i':
			interactive = 1;
			break;
		default:
			warnx("unsupported option %c", c);
			print_help();
		}
	}

	/* Report extra unparsed arguments. */
	if (optind < argc) {
		warnx("bogus arguments:");
		while (optind < argc)
			fprintf(stderr, "%s ", argv[optind++]);
		fputc('\n', stderr);
		exit(1);
	}
	if (!(dpy = XOpenDisplay(display)))
		errx(1, "can't open display %s", display);
	set_close_on_exec(ConnectionNumber(dpy));

	/* forked commands should not get X console tty as their stdin */
	dup2(open("/dev/null", O_RDONLY), STDIN_FILENO);

	/* Set our own specific Atoms. */
	rp_selection = XInternAtom(dpy, "RP_SELECTION", False);

	/* TEXT atoms */
	xa_string = XA_STRING;
	xa_compound_text = XInternAtom(dpy, "COMPOUND_TEXT", False);
	xa_utf8_string = XInternAtom(dpy, "UTF8_STRING", False);

	init_control_socket_path();

	if (cmd_count > 0) {
		int j, exit_status = 0;

		for (j = 0; j < cmd_count; j++) {
			if (!send_command(interactive, (unsigned char *)cmd[j]))
				exit_status = 1;
			free(cmd[j]);
		}

		free(cmd);
		XCloseDisplay(dpy);
		return exit_status;
	}

	/* For child processes to know */
	snprintf(pid, sizeof(pid), "%d", getpid());
	setenv("SDORFEHS_PID", pid, 1);

	/* Must be first */
	register_atom(&_net_supported, "_NET_SUPPORTED");

	register_atom(&wm_change_state, "WM_CHANGE_STATE");
	register_atom(&wm_colormaps, "WM_COLORMAP_WINDOWS");
	register_atom(&wm_delete, "WM_DELETE_WINDOW");
	register_atom(&wm_name, "WM_NAME");
	register_atom(&wm_protocols, "WM_PROTOCOLS");
	register_atom(&wm_state, "WM_STATE");
	register_atom(&wm_take_focus, "WM_TAKE_FOCUS");

	register_atom(&_net_active_window, "_NET_ACTIVE_WINDOW");
	register_atom(&_net_client_list, "_NET_CLIENT_LIST");
	register_atom(&_net_client_list_stacking, "_NET_CLIENT_LIST_STACKING");
	register_atom(&_net_current_desktop, "_NET_CURRENT_DESKTOP");
	register_atom(&_net_number_of_desktops, "_NET_NUMBER_OF_DESKTOPS");
	register_atom(&_net_workarea, "_NET_WORKAREA");

	register_atom(&_net_wm_name, "_NET_WM_NAME");
	register_atom(&_net_wm_pid, "_NET_WM_PID");
	register_atom(&_net_wm_state, "_NET_WM_STATE");
	register_atom(&_net_wm_state_fullscreen, "_NET_WM_STATE_FULLSCREEN");
	register_atom(&_net_wm_window_type, "_NET_WM_WINDOW_TYPE");
	register_atom(&_net_wm_window_type_dialog, "_NET_WM_WINDOW_TYPE_DIALOG");
	register_atom(&_net_wm_window_type_dock, "_NET_WM_WINDOW_TYPE_DOCK");
	register_atom(&_net_wm_window_type_splash, "_NET_WM_WINDOW_TYPE_SPLASH");
	register_atom(&_net_wm_window_type_tooltip,
	    "_NET_WM_WINDOW_TYPE_TOOLTIP");
	register_atom(&_net_wm_window_type_utility,
	    "_NET_WM_WINDOW_TYPE_UTILITY");
	register_atom(&_net_supporting_wm_check,
	    "_NET_SUPPORTING_WM_CHECK");

	/* Setup signal handlers. */
	XSetErrorHandler(handler);
	set_sig_handler(SIGALRM, alrm_handler);
	set_sig_handler(SIGTERM, sighandler);
	set_sig_handler(SIGINT, sighandler);
	set_sig_handler(SIGHUP, hup_handler);
	set_sig_handler(SIGCHLD, chld_handler);

	if (bar_mkfifo() == -1)
		return 1;

	/* Setup our internal structures */
	init_defaults();
	init_window_stuff();
	init_xrandr();
	init_screens();

	update_modifier_map();
	init_user_commands();
	initialize_default_keybindings();
	history_load();
	init_bar();

	scanwins();

	c = read_startup_files(alt_rcfile);
	if (c == -1)
		return 1;
	if (c == 0) {
		/* No config file, just do something basic. */
		cmdret *result;
		if ((result = command(0, "hsplit")))
			cmdret_free(result);
	}

	/* Indicate to the user that we have booted. */
	if (defaults.startup_message)
		show_welcome_message();

	/* If no window has focus, give the key_window focus. */
	if (current_window() == NULL)
		set_window_focus(rp_current_screen->key_window);

#ifdef __OpenBSD__
	/* cpath/wpath/fattr needed for history_save() */
	pledge("stdio rpath cpath wpath fattr unix proc exec", NULL);
#endif

	listen_for_commands();
	listen_for_events();

	return 0;
}
